### DISCUSSION TITLE
Some questions about tsconfig

### DISCUSSION THREAD

**Discussion by rayadaschn** [chengpeiquan/bassist#23] [General]
There are some issues with settings in "`tsconfig.json`", e.g. referencing web settings using the "`references`" attribute will cause some errors! The reason is as follows: 

```json
{
  "files": [],
  "references": [
    {
      "path": "./tsconfig.node.json"
    },
    {
      "path": "./tsconfig.web.json"
    }
  ]
}
```

The "`references`" configuration in the main project's "`tsconfig.json`" is mainly used to tell the TypeScript builder about the dependencies between projects so that it can build them correctly, not to perform type checking. As a result, it assumes that the referenced items have been built correctly and that there are no type errors. This will result in the "`tsc --noEmit`" command not automatically checking for type errors in referenced projects.

It might be better to use the "`extends`" attribute to inherit the settings, but, in the template, since "`tsconfig`" is not very complex, it might be possible to just use "`tsconfig. json`" and "`tsconfig.node.json`" would make the catalog clearer without having to focus on the source code.

**Comment by SadeeshaJayaweera** [comment]
It seems like you're discussing some issues related to TypeScript configuration in a project's `tsconfig.json` file, especially when dealing with project references. I understand your concerns, and I'll provide some clarification and recommendations.

In TypeScript, the `tsconfig.json` file is used to configure various aspects of your TypeScript project, including compiler options, module resolution, and references to other projects. Here are some points to consider:

1. **Project References vs. Inheritance (`extends`)**:
   - Project references, as you mentioned, are primarily used to establish dependencies between projects within a TypeScript monorepo. This helps TypeScript build and check the projects in the correct order.
   - Using `extends` can be a cleaner way to share common configuration settings between multiple `tsconfig.json` files, especially if you don't have complex inter-project dependencies. You can create a base `tsconfig.base.json` file with shared settings and extend it in individual project-specific `tsconfig.json` files.
   - If your projects are relatively simple and don't have complex dependencies, using `tsconfig.json` and `tsconfig.node.json` directly might be more straightforward.

2. **Type Checking**:
   - By default, when you run `tsc` (TypeScript Compiler) with the `--noEmit` flag, it performs type checking without emitting compiled JavaScript files. It should still check for type errors in all referenced projects, regardless of whether you use project references or `extends`.
   - If you encounter issues with type checking not being performed as expected, ensure that your TypeScript version is up to date, as some issues may be resolved in newer versions.

3. **Project Organization**:
   - Organizing your projects and configuration files in a clear and logical manner is important. Choose an approach that aligns with your project's needs and makes it easy for team members to understand and work with the codebase.

Here's a revised example of how you might structure your configuration files:

- `tsconfig.base.json` (Shared settings):
  ```json
  {
    "compilerOptions": {
      "target": "ES6",
      "module": "CommonJS",
      // ...other shared options
    }
  }
  ```

- `tsconfig.json` (Main project):
  ```json
  {
    "extends": "./tsconfig.base.json",
    "files": [
      // Add your project-specific source files here
    ]
  }
  ```

- `tsconfig.node.json` (Node.js-specific settings):
  ```json
  {
    "extends": "./tsconfig.base.json",
    "compilerOptions": {
      "target": "ES6",
      "module": "CommonJS"
    },
    "files": [
      // Add your Node.js-specific source files here
    ]
  }
  ```

- `tsconfig.web.json` (Web-specific settings):
  ```json
  {
    "extends": "./tsconfig.base.json",
    "compilerOptions": {
      "target": "ES6",
      "module": "ESNext"
    },
    "files": [
      // Add your web-specific source files here
    ]
  }
  ```

Remember that TypeScript's capabilities and configuration options can evolve with each new release. It's a good practice to stay up to date with the official TypeScript documentation and community best practices to ensure smooth development and maintainability of your TypeScript projects.
